File: leetcode/216_combination_sum_iii.py
class Solution:
    # @param {integer} k
    # @param {integer} n
    # @return {integer[][]}
    def dfs(self, start, cur, rest):
      if (len(cur) > self.k or rest < 0):
        return

      if (len(cur) == self.k and rest == 0):
        self.res.append(cur)
      else:
        for i in range(start, 10):
          self.dfs(i+1, cur+[i], rest-i)

    def combinationSum3(self, k, n):
      self.res = []
      self.n = n
      self.k = k
      self.dfs(1, [], n)
      return self.res


s = Solution()
k = 3
n = 28
print s.combinationSum3(k, n)

File: leetcode/216_combination_sum_iii.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function(k, n) {
  var res = [];
  (function dfs(start, cur, rest){
    if (cur.length === k && rest === 0) {
      res.push(cur.slice(0));
    } else {
      if (cur.length < k) {
        for (var i = start; i < 10; i += 1) {
          if (rest - i >= 0) {
            cur.push(i);
            dfs(i+1, cur, rest-i);
            cur.pop();
          } else {
            break;
          }
        }
      }
    }

  }(1, [], n));

  return res;
};


console.log(combinationSum3(3, 9));
